File: day-07/day.rb
def run
  input = parse_input(read_input)

  # code to run part 1 and part 2
  puts min_alignment_cost(input, :linear)
  puts min_alignment_cost(input, :triangular)
end

def read_input(filename = "input.txt")
  if !STDIN.tty?
    ARGF.read
  else
    filename = File.expand_path(ARGV[0] || filename, __dir__)
    File.read(filename)
  end
end

def parse_input(input)
  input.split(",").map(&:to_i)
end

### CODE HERE ###
FUEL_STRATEGIES = {
  linear: ->(p, i) { (p - i).abs },
  triangular: ->(p, i) { n = (p - i).abs; n * (n + 1) / 2 }
}

def min_alignment_cost(positions, strategy)
  min, max = positions.minmax
  strategy = FUEL_STRATEGIES.fetch(strategy)
  (min..max).map do |i|
    positions.sum { |p| strategy.call(p, i) }
  end.min
end

### TESTS HERE ###
require "rspec"

describe "day" do
  let(:input) do
    parse_input(File.read("test.txt"))
  end

  let(:actual_input) do
    parse_input(File.read("input.txt"))
  end

  it "should solve part 1" do
    expect(min_alignment_cost(input, :linear)).to eq 37
    expect(min_alignment_cost(actual_input, :linear)).to eq 341558
  end

  it "should solve part 2" do
    expect(min_alignment_cost(input, :triangular)).to eq 168
    expect(min_alignment_cost(actual_input, :triangular)).to eq 93214037
  end
end

run if $PROGRAM_NAME == __FILE__ || $PROGRAM_NAME.end_with?("ruby-memory-profiler")

File: day-07/day.rs
use std::env;
use std::fs::File;
use std::io::BufReader;
use std::io::BufRead;
//use std::str::FromStr;
//use std::convert::Infallible;

fn main() {
    let args: Vec<String> = env::args().collect();
    let mut filename = "input.txt";
    if args.len() > 1 {
        filename = &args[1];
    }
    let file = File::open(filename).expect("input file to be found");
    let input = &parse_input(read_input(file));

    println!("{}", min_alignment_cost("linear", input));
    println!("{}", min_alignment_cost("triangular", input));
}

fn read_input(file: File) -> Vec<String> {
    let reader = BufReader::new(file);
    reader.lines().map(|l| l.unwrap()).collect()
}

fn parse_input(input: Vec<String>) -> Vec<isize> {
    input[0].split(",").map(|l| {
        // replace 'String' with the type we're trying to parse
        l.parse().unwrap()
    }).collect()
}

fn min_alignment_cost(strategy: &str, positions: &Vec<isize>) -> isize {
    let (min, max) = positions[1..].iter().fold((positions[0], positions[0]), |(mut min, mut max), x| {
        if *x < min { min = *x; }
        if *x > max { max = *x; }
        (min, max)
    });

    let fuel_calc = fuel_strategy(strategy);
    let mut min_fuel = None;
    for i in min..max {
        let s: isize = positions.iter().map(|p| fuel_calc(*p, i) ).sum();
        match min_fuel {
            None => min_fuel = Some(s),
            Some(m) => {
                if s < m { min_fuel = Some(s); }
            }
        }
    }
    min_fuel.unwrap()
}

fn fuel_strategy(s: &str) -> impl Fn(isize, isize) -> isize {
    if s == "linear" {
        |p: isize, i: isize| { (p - i).abs() }
    } else if s == "triangular" {
        |p: isize, i: isize| { let n = (p - i).abs(); n * (n + 1) / 2 }
    } else {
        panic!("unsupported strategy {}", s)
    }
}

#[cfg(test)]
mod aoc_tests {
    use super::*;

    fn test_input() -> Vec<isize> {
        let filename = "test.txt";
        let file = File::open(filename).expect("input file to be found");
        parse_input(read_input(file))
    }

    fn actual_input() -> Vec<isize> {
        let filename = "input.txt";
        let file = File::open(filename).expect("input file to be found");
        parse_input(read_input(file))
    }

    #[test]
    fn test_part1() {
        assert_eq!(min_alignment_cost("linear", &test_input()), 37);
        assert_eq!(min_alignment_cost("linear", &actual_input()), 341558);
    }

    #[test]
    fn test_part2() {
        assert_eq!(min_alignment_cost("triangular", &test_input()), 168);
        assert_eq!(min_alignment_cost("triangular", &actual_input()), 93214037);
    }
}
